### ISSUE TITLE
[Gap Analysis] Implement Unified Model Evaluation Framework and Learning Curves

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#334]
### User Story

> As a machine learning engineer, I want a comprehensive and unified model evaluation framework, including tools for generating learning curves and analyzing bias-variance trade-offs, so that I can thoroughly assess model performance, diagnose issues, and make informed decisions about model selection and improvement.

---

### **Problem: Lack of a Unified Model Evaluation Framework and Advanced Techniques**

The `src/Evaluation` module currently offers only a `DefaultModelEvaluator.cs`, which likely provides basic functionality. While individual metrics are being addressed in `src/Validation`, there is a significant gap in providing a comprehensive, unified evaluation framework that leverages these metrics and offers higher-level evaluation techniques crucial for robust model assessment.

---

### **Phase 1: Implement a Unified Model Evaluation Report/Interface**

**Goal:** Create a central component to perform and report on various model evaluations.

#### **AC 1.1: Create `ModelEvaluator.cs` (18 points)**
**Requirement:** Implement a class to orchestrate model evaluation and generate reports.

-   [ ] **File:** `src/Evaluation/ModelEvaluator.cs`
-   [ ] **Class:** `public class ModelEvaluator<T>`
-   [ ] **Constructor:** Takes an `IModel<T>` and a `TaskType` (e.g., Classification, Regression).
-   [ ] **Methods:**
    -   `public EvaluationReport Evaluate(Matrix<T> X, Vector<T> y)`: Runs a full evaluation and returns a report object.
    -   `public void PrintReport(EvaluationReport report)`: Prints a human-readable summary.
-   [ ] **Logic:** Internally uses the metrics implemented in `src/Validation/Metrics` to compute relevant scores based on `TaskType`.

#### **AC 1.2: Create `EvaluationReport.cs` (8 points)**
**Requirement:** Define a data structure to hold evaluation results.

-   [ ] **File:** `src/Evaluation/EvaluationReport.cs`
-   [ ] **Class:** `public class EvaluationReport`
-   [ ] **Properties:** Stores common metrics (e.g., Accuracy, Precision, Recall, F1, AUC for classification; MAE, MSE, RMSE, R2 for regression).

#### **AC 1.3: Unit Tests for `ModelEvaluator` and `EvaluationReport` (10 points)**
**Requirement:** Verify the functionality of the evaluator and report generation.

-   [ ] **File:** `tests/UnitTests/Evaluation/ModelEvaluatorTests.cs`
-   [ ] **Test Cases:** Test `Evaluate` with a simple model and dataset, ensuring correct metrics are calculated and stored in the report.

---

### **Phase 2: Implement Learning Curve Analysis**

**Goal:** Provide tools to generate and analyze learning curves for bias-variance diagnosis.

#### **AC 2.1: Create `LearningCurveAnalyzer.cs` (18 points)**
**Requirement:** Implement a utility to generate learning curve data.

-   [ ] **File:** `src/Evaluation/LearningCurveAnalyzer.cs`
-   [ ] **Class:** `public static class LearningCurveAnalyzer<T>`
-   [ ] **Method:** `public static LearningCurveData GenerateLearningCurve(IModel<T> model, Matrix<T> X, Vector<T> y, IEnumerable<double> trainSizes, IMetric<T> metric, int nSplits)`.
-   [ ] **Logic:** Trains the model on increasing subsets of the training data and evaluates performance on both training and validation sets, returning `LearningCurveData`.

#### **AC 2.2: Create `LearningCurveData.cs` (5 points)**
**Requirement:** Define a data structure to hold learning curve results.

-   [ ] **File:** `src/Evaluation/LearningCurveData.cs`
-   [ ] **Class:** `public class LearningCurveData`
-   [ ] **Properties:** Stores `trainSizes`, `trainScores`, `validationScores`.

#### **AC 2.3: Unit Tests for `LearningCurveAnalyzer` (10 points)**
**Requirement:** Verify the correctness of learning curve generation.

-   [ ] **File:** `tests/UnitTests/Evaluation/LearningCurveAnalyzerTests.cs`
-   [ ] **Test Cases:** Test `GenerateLearningCurve` with a simple model and dataset, ensuring correct data points are generated for train and validation scores.

---

### **Definition of Done**

-   [ ] All checklist items are complete.
-   [ ] `ModelEvaluator` and `EvaluationReport` are implemented and unit-tested.
-   [ ] `LearningCurveAnalyzer` and `LearningCurveData` are implemented and unit-tested.
-   [ ] All new tests pass.
---

## ⚠️ CRITICAL ARCHITECTURAL REQUIREMENTS

**Before implementing this user story, you MUST review:**
- **📋 Full Requirements:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
- **📐 Project Rules:** [`.github/PROJECT_RULES.md`](../.github/PROJECT_RULES.md)

#### 1. INumericOperations<T> Usage (CRITICAL)
- [ ] Include `protected static readonly INumericOperations<T> NumOps = MathHelper.GetNumericOperations<T>();` in base class
- [ ] NEVER hardcode `double`, `float`, or specific numeric types - use generic `T`
- [ ] NEVER use `default(T)` - use `NumOps.Zero` instead
- [ ] Use `NumOps.Zero`, `NumOps.One`, `NumOps.FromDouble()` for values
- [ ] Use `NumOps.Add()`, `NumOps.Multiply()`, etc. for arithmetic
- [ ] Use `NumOps.LessThan()`, `NumOps.GreaterThan()`, etc. for comparisons

#### 2. Inheritance Pattern (REQUIRED)
- [ ] Create `I{FeatureName}.cs` in `src/Interfaces/` (root level, NOT subfolders)
- [ ] Create `{FeatureName}Base.cs` in `src/{FeatureArea}/` inheriting from interface
- [ ] Create concrete classes inheriting from Base class (NOT directly from interface)

#### 3. PredictionModelBuilder Integration (REQUIRED)
- [ ] Add private field: `private I{FeatureName}<T>? _{featureName};` to `PredictionModelBuilder.cs`
- [ ] Add Configure method taking ONLY interface (no parameters):
  ```csharp
  public IPredictionModelBuilder<T, TInput, TOutput> Configure{FeatureName}(I{FeatureName}<T> {featureName})
  {
      _{featureName} = {featureName};
      return this;
  }
  ```
- [ ] Use feature in `Build()` with default: `var {featureName} = _{featureName} ?? new Default{FeatureName}<T>();`
- [ ] Verify feature is ACTUALLY USED in execution flow

#### 4. Beginner-Friendly Defaults (REQUIRED)
- [ ] Constructor parameters with defaults from research/industry standards
- [ ] Document WHY each default was chosen (cite papers/standards)
- [ ] Validate parameters and throw `ArgumentException` for invalid values

#### 5. Property Initialization (CRITICAL)
- [ ] NEVER use `default!` operator
- [ ] String properties: `= string.Empty;`
- [ ] Collections: `= new List<T>();` or `= new Vector<T>(0);`
- [ ] Numeric properties: appropriate default or `NumOps.Zero`

#### 6. Class Organization (REQUIRED)
- [ ] One class/enum/interface per file
- [ ] ALL interfaces in `src/Interfaces/` (root level)
- [ ] Namespace mirrors folder structure (e.g., `src/Regularization/` → `namespace AiDotNet.Regularization`)

#### 7. Documentation (REQUIRED)
- [ ] XML documentation for all public members
- [ ] `<b>For Beginners:</b>` sections with analogies and examples
- [ ] Document all `<param>`, `<returns>`, `<exception>` tags
- [ ] Explain default value choices

#### 8. Testing (REQUIRED)
- [ ] Minimum 80% code coverage
- [ ] Test with multiple numeric types (double, float)
- [ ] Test default values are applied correctly
- [ ] Test edge cases and exceptions
- [ ] Integration tests for PredictionModelBuilder usage

---

**⚠️ Failure to follow these requirements will result in repeated implementation mistakes and PR rejections.**

**See full details:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
